### DISCUSSION TITLE
Tuns counter for winding in h4

### DISCUSSION THREAD

**Discussion by tektonandroid** [kachurovskiy/nanoels#231] [Ideas]
Good afternoon everyone. I would like to use my lathe to count the turns when winding transformers. Would it be possible to implement a turns counter in the H4 software?
Regards and thanks

**Comment by tektonandroid** [comment]
Estuve estudiando el asunto, pero yo no sé nada de programar...he conseguido algo de ayuda a traves de buscadores y he deducido lo siguiente:
Debo declarar unas variables nuevas:
bool showTurns = false //Indica si debe de verse en el display.
bool resetSpindleTurns = false //la activación de esta variable pone spindleTurns a 0. Debería de poder ejecutarla con el mismo botón que pone a 0 los angulos del husillo (creo que manteniendo pulsado el stop), aunque tampoco sé cómo se implementa.
long spindleTurns = 0 //Esta variable indica el número a mostrar. Debería de poder mostrar números negativos y positivos para sumar o restar vueltas de cable. ¿De donde puede sacar el dato necesario? 
Se me ha sugerido que, para contar y descontar, modifique la función IRAM ATTR de la siguiente forma (marco con * las lineas añadidas):

void IRAM_ATTR onAsyncTimer() {
  Axis* a = getAsyncAxis();
  if (!isOn || a->movingManually || (mode != MODE_ASYNC && mode != MODE_A1)) {
    return;
  } else if (dupr > 0 && a->pos < a->leftStop) {
    if (a->pos <= a->motorPos) {
      a->pos++;
    }
    a->motorPos++;
    a->posGlobal++;
  *spindleTurns++; // Incrementar el contador
  } else if (dupr < 0 && a->pos > a->rightStop) {
    if (a->pos >= a->motorPos + a->backlashSteps) {
      a->pos--;
    }
    a->motorPos--;
    a->posGlobal--;
    *spindleTurns--; // Decrementar el contador
  } else {
    return;
  }

Además, se me sugiere que alterne las RPM con los GRADOS del husillo y las Vueltas que quiero mostrar, de la siguiente forma:
 void buttonDisplayPress() 
{ if (!showAngle && !showTacho && !showTurns) {
	 showAngle = true;
 } else if (showAngle) {
	showAngle = false;
	showTacho =true;
	showTurns = false; 
 } else if (showTacho) {
	showTacho = false;
	showTurns = true;
	showAngle = false; 
 } else if (showTurns) {
	showTacho = false;
	showTurns = false;
	showAngle = true; 
 } else { showTurns = false;
	showTacho = false;
	}
 }

También se me sugiere que incluya en la función updateDisplay lo siguiente:
if (showTurns) {
  lcd.setCursor(0, 3);
  lcd.print("Turns: ");
  lcd.print(spindleTurns);
  printLcdSpaces(charIndex);
}
Pero en esa función hay demasiadas condiciones y no sabría donde incluir esta parte,,,,
No sé si sirve de algo....Gracias a todos. Espero un poco de ayuda (mucha, en realidad), ya que entiendo poco de todo este código....

**Comment by kachurovskiy** [comment]
Try adding 

```
      charIndex += lcd.print((float) abs(spindlePos) / ENCODER_STEPS_INT, 2);
      charIndex += lcd.print(" ");
```

after the line 1005 printing the "Angle " string - https://github.com/kachurovskiy/nanoels/blob/main/h4/h4.ino#L1005 - this will show the number of turns with 2 precision signs in the Angle display mode - and the degree angle would be shown after.

Reset angle and turns with the Stop button.

**Reply by tektonandroid** [reply]
Dios mío.. me pase 2 horas inventando código y preguntando a IA, y me lo has resuelto con dos líneas y mucho conocimiento... BRAVO Y ENHORABUENA...y 100000 gracias

**Comment by ilmarietto123** [comment]
Hi @tektonandroid,
I modified too the code starting from the @kachurovskiy suggestion to have more digits for the turns count.
If you are interested, you can replace [lines from 1005 to 1007](https://github.com/kachurovskiy/nanoels/blob/8cc4b13d93870869a071ace68bb6aeda617e1e79/h4/h4.ino#L1005-L1007) with this code:
```
charIndex += lcd.print(char(224));  //  print "alpha" char instead of "Angle " to spare more digit for turns...
charIndex += lcd.print(":");        //  ... because if the line is 21 char, the 21st char overwrites the 1st top left of the display
float angleValue = 360.0 * spindleModulo(spindlePos) / ENCODER_STEPS_INT;
charIndex += lcd.print(angleValue > 99.99 ? "" : (angleValue > 9.99) ? " " : "  ");   // adding spaces before angle, so that comma stays in place
charIndex += lcd.print(angleValue, 2);
charIndex += lcd.print(char(223));
charIndex += lcd.print(" T:");                                          //  "T:" for turns  
charIndex += lcd.print((float)abs(spindlePos) / ENCODER_STEPS_INT, 1);  //  shows only 1 decimal because if precision is needed, better look at the angle beside
```

The results is a line like this:
"α:XXX.XX° T:YYYYYY.Y"
Theoretically you can count quite 1 million turns before messing the display (999999.9 turns).

**Reply by tektonandroid** [reply]
Fantástico. Lo aplicaré en mi torno también.
Saludos.
